cooked-validators/src/Cooked/Attack.hs

```diff
@@ -0,0 +1,158 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cooked.Attack where
+
+import Cooked.MockChain
+import Cooked.Tx.Constraints
+import Cooked.Tx.Constraints.Optics
+import Data.Maybe
+import qualified Ledger as L hiding (validatorHash)
+import qualified Ledger.Typed.Scripts as L
+import Optics.Core
+import qualified PlutusTx as Pl
+import qualified PlutusTx.Numeric as Pl (negate)
+import Type.Reflection
+
+-- * The type of attacks, and functions to turn optics into attacks
+
+-- | The type of attacks that operate on a single transaction. The idea is to
+-- try to modify a transaction, or return @Nothing@ if the modification does not
+-- apply; use this with the 'somewhere' and 'everywhere' modalities from
+-- 'Cooked.MockChain.Monad'.
+type Attack = TxSkel -> Maybe TxSkel
+
+-- | The simplest way to make an attack from an optic: Try to apply the given
+-- function to all of the optic's foci. Fail, returning @Nothing@, if nothing is
+-- in focus or if all foci are evaluated to @Nothing@.
+mkAttack :: Is k A_Traversal => Optic' k is TxSkel a -> (a -> Maybe a) -> Attack
+mkAttack optic f skel =
+  let (skel', modified) =
+        mapAccumROf
+          optic
+          ( \flag x -> case f x of
+              Just y -> (y, True)
+              Nothing -> (x, flag)
+          )
+          False
+          skel
+   in if modified then Just skel' else Nothing
+
+-- * General helpers
+
+-- | Add a label to a 'TxSkel'. If there is already a pre-existing label, the
+-- given label will be added, forming a pair @(newlabel, oldlabel)@.
+addLabel :: LabelConstrs x => x -> TxSkel -> TxSkel
+addLabel newlabel =
+  over
+    txLabelL
+    ( \case
+        TxLabel Nothing -> TxLabel $ Just newlabel
+        TxLabel (Just oldlabel) -> TxLabel $ Just (newlabel, oldlabel)
+    )
+
+-- | try to use the given function to modify the elements of the given list. If
+-- at least one modification is successful, return 'Just' the modified list,
+-- otherwise fail returning 'Nothing'.
+someJust :: (a -> Maybe a) -> [a] -> Maybe [a]
```

Comment:
nit-picking: it seems like it this is used only in `dupTokenAttack`, if that's really the case, maybe we move this into the `where` clause there?

Reply:
Yes, it's only used there, but I imagined other attacks that we'll write in the future might have a use for it or something similar. We used almost the same function for the `datumHijackingAttack`, before I refactored it. Butmaybe you're right and it's better not to have too many cryptic helper functions.